modules/host/update.go

```diff
@@ -313,7 +295,25 @@ func (h *Host) ProcessConsensusChange(cc modules.ConsensusChange) {
 		h.log.Println(err)
 	}
 
-	// Handle the list of action items.
+	// Handle the list of action items. Action items require host locks and
+	// potentially require waiting for long periods of time while various
+	// network communications finish. To prevent the host lock from being held
+	// while the network I/O is blocking the action item execution, the action
+	// items are handled in their own thread. A wait group is used to prevent
+	// the host from releasing ThreadGroup control until all threads that this
+	// function have spawned have completed.
+	wg := new(sync.WaitGroup)
+	// The host should not release thread group control until all spawned
+	// threads have completed execution, though waiting for them should not
+	// block this function. The host should also not release thread group
+	// control until this function has terminated, hence the use of 'defer
+	// func() { go func() }' to coordinate waiting and done-ing.
+	defer func() {
+		go func() {
+			wg.Wait()
```

Comment:
It's not unbounded, because you will only spawn a thread to handle an action item if there is an action item, and since the host is locked you know that action items won't be getting added while you are spinning up the threads.

There is a call to `tg.Add()` at the beginning of `ProcessConsensusChange`, which means all future calls will be blocked. It's important for `Flush()` to wait until all action items have been handled, otherwise you aren't going to know (during testing) if the storage proof has been submitted yet. The purpose of `Flush()` is definitely to make it easy to know that certain actions have completed.

I could potentially see issues if some 'attacker' has formed 10,000 file contracts with you that all expire in the same, but they pay a small price for each contract they form, and that's an attack vector regardless of how the threading is operating.

So basically, every action item that you have to deal with this block would have been created prior to the call to `ProcessConsensusChange` and would mean at least a little bit of income each.

***if it's not clear, `Flush()` and `Stop()` will block all incoming calls to `tg.Add()` which occurs at the top of the `ProcessConsensusChange` function. So even if the call to `ProcessConsensusChange` completes, you know that there will not be another one to add more action items.